pkg/plugins/generic/generic_plugin.go

```diff
@@ -69,31 +111,41 @@ func (p *GenericPlugin) OnNodeStateChange(new *sriovnetworkv1.SriovNetworkNodeSt
 	p.DesireState = new
 
 	needDrain = needDrainNode(new.Spec.Interfaces, new.Status.Interfaces)
-	needReboot = needRebootNode(new, &p.LoadVfioDriver, &p.LoadVirtioVdpaDriver)
+	needReboot = needRebootNode(new, p.DriverStateMap)
 
 	if needReboot {
 		needDrain = true
 	}
 	return
 }
 
+// ////////////// for testing purposes only ///////////////////////
+func (p *GenericPlugin) GetDriverStateMap() DriverStateMapType {
+	return p.DriverStateMap
+}
+
+func (p *GenericPlugin) TestDriverLoading(state *sriovnetworkv1.SriovNetworkNodeState) {
+	for _, driverState := range p.DriverStateMap {
+		if !driverState.DriverLoaded && driverState.NeedDriverFunc(state, driverState) {
+			driverState.DriverLoaded = true
+		}
+	}
+}
+
+//////////////////////////////////////////////////////////////////
+
 // Apply config change
 func (p *GenericPlugin) Apply() error {
 	glog.Infof("generic-plugin Apply(): desiredState=%v", p.DesireState.Spec)
-	if p.LoadVfioDriver == loading {
-		if err := p.HostManager.LoadKernelModule("vfio_pci"); err != nil {
-			glog.Errorf("generic-plugin Apply(): fail to load vfio_pci kmod: %v", err)
-			return err
-		}
-		p.LoadVfioDriver = loaded
-	}
 
-	if p.LoadVirtioVdpaDriver == loading {
-		if err := p.HostManager.LoadKernelModule("virtio_vdpa"); err != nil {
-			glog.Errorf("generic-plugin Apply(): fail to load virtio_vdpa kmod: %v", err)
-			return err
+	for _, driverState := range p.DriverStateMap {
```

Comment:
can you move this to a helper method: `SyncDriverState()` and call it after L157 while at it ?

flow would be clearer IMO